feature(게시글 상세 조회): 댓글 컴포넌트 기능 구현

Derek-94 opened the pull request:
## 상세내용

- @hyunju95 님이 만드신 #9 의 기능 구현.
- `container component` 개발.

## 데이터의 흐름 설명
[약간 어려울수 있으니, 자세히 설명 하겠습니다.]

### 1. **결론부터)** 최종적으로 그리는 것은, `CommentContainer` 컴포넌트 입니다.

![image](https://user-images.githubusercontent.com/52649378/115485355-7dd3fd00-a28f-11eb-97d5-54e6461abe57.png)

상단 그림에 보면 `CommentContainer` 를 그리는 모습입니다.

### 2. `CommentContainer` 는  presentational인 `CommentArea` 컴포넌트에게 **3가지를 인자**를 넘깁니다.

![image](https://user-images.githubusercontent.com/52649378/115485289-5e3cd480-a28f-11eb-8cb6-d37db187152c.png)

  - `댓글 등록시 수행하는 이벤트 함수`
  - `좋아요 눌렀을때 수행되는 이벤트 함수`
  - `dummyData`

이렇게 세가지를 넘깁니다. 

### 3. `CommentArea` 는 이 세 가지 인자를 `CommentAreaProps` 라는 interface로 묶어서 전달 받습니다. 

![image](https://user-images.githubusercontent.com/52649378/115485446-b4aa1300-a28f-11eb-80b9-752fadfd9890.png)

  - `typescript` 때문에 이렇게 넘겨줘야하는 것으로 판단됩니다.

### 4. `CommentArea` 에서는 크게 다음 기능을 합니다.

  - `form` 태그의 `onSubmit` 기능 담당.
  - `input` 태그의 `onChange` 기능 담당.
  - 전달받은 `dummyData` 를 `Comment` 컴포넌트로 렌더링.

### 5. 마지막으로 전달받은 `dummyData` 를 배열로 map을 돌면서 `Comment` 컴포넌트르 렌더링합니다.

---

## 고민사항
`CommentAreaProps` 라는 interface를 어디다가 저장을 해야할지 고민입니다. 

이 사항 관련해서 꼭 **같이** 고민해봤으면 좋겠어요.

## +α) 참고사항
### 이번 PR에서는 `useState` 라는 react hooks를 사용했습니다. 
정말정말 쉬운 개념이니, 모르신다면 꼭 [여기](https://ko.reactjs.org/docs/hooks-overview.html)를 참고해주세요. 제발 꼭!

closes #10

Derek-94 committed:
feat(comment): 댓글 컴포넌트 컨테이너 개발 시작.

- 댓글 컨테이너 구조 설계.
- api 통신 결과를 comment component에 전달 예정.

Derek-94 committed:
feat(comment): 댓글 컴포넌트 컨테이너 개발 시작.

- 댓글 컨테이너 구조 설계.
- api 통신 결과를 comment component에 전달 예정.

Derek-94 committed:
feat(comment): 댓글 컨테이너 컴포넌트 개발.

- 댓글 presentational 컴포넌트에게 필요한 이벤트 핸들링 함수 전달.
- api가 준비가 안되어 useState 활용함.

Derek-94 committed:
feat(comment): 댓글 컨테이너 컴포넌트 개발.

- 댓글 presentational 컴포넌트에게 필요한 이벤트 핸들링 함수 전달.
- api가 준비가 안되어 useState 활용함.

Derek-94 committed:
chore: eslint랑 prettier 적용.

- DummyData 형식 수정.

Derek-94 committed:
chore: eslint랑 prettier 적용.

- DummyData 형식 수정.

hyunju95 committed:
feat(article/comment) Add Comment Area Component

hyunju95 committed:
feat(article/comment) : Add Component Area

Derek-94 committed:
feature(comment): 댓글 컴포넌트 기능 통합 개발.

- presenter component와 container component 병합작업.
- dummyData를 추가되어 하단에 댓글이 추가되는 기능 추가.

Derek-94 committed:
feat(comment): 댓글 컴포넌트 추가개발

- 변수 이름 수정.
- 댓글 갯수를 동적으로 렌더링할 수 있게 수정.

Derek-94 committed:
Merge branch 'feature/10' of https://github.com/dev-floor/untitled-client into feature/10

hyunju95 commented on src/Article/container/CommentContainer.tsx:
```diff
@@ -0,0 +1,29 @@
+    const [comments, setComments] = useState(dummyData);
+
+    const onPressFavorite = (targetComment: ArticleCommentApi) => {
+        setComments(commentList => commentList.map(comment =>
```
이 부분에서 commentList를 전부 map도는 방식 말고 다른 방법도 있나 찾아보면 좋을것같아요. 우리프로젝트에서는 물론 댓글이 많아야 몇백개여서 큰 문제가 되지않을것 같지만 혹시 나중에 성능상 문제가 생길수도있으니? 시간되면 효율적으로 comment 찾는방법이있을지 찾아보면 좋을듯합니당

hyunju95 replied:
추가적으로 setComments 요 구문 추가설명가능할까요 어디에 선언된형태인건지? 아니면 저렇게 원래 지정해서 사용해야하는 구문인건지? 궁금합니닷

hyunju95 replied:
저게 상태변수 그건가요?

Derek-94 replied:
음.. 일단은 생각나는게 없더라구요 ㅜ `api` 가 구축되기전에는 이렇게라도 개발하자, 해서 이런 방향으로 했어요.
추후에 `api` 가 뚫려도.. 다른 방법이 있으려나 모르겠네요..! 
콜백 함수 인자로 받은 댓글 객체 하나를 댓글 목록에서 찾는 것 외에는.. 음 이건 고민을 같이해봐야할것 같아요.

다음 질문을 답변드리자면, `setComments` 라는 함수는 상단에 보시면 `useState` 로 선언된 함수에요.

7번째 줄에서 `useState` 를 사용하면서 인자를 `dummyData` 를 던저주면, 첫번째 인자인 `comments`는 초기값이
`dummyData` 로 설정이 됩니다. 

그 다음부터는 `comments` 를 수정하고 싶다면 두번째 인자인 `setComments` 를 사용해야합니다! 
이 `setComments` 를 통해서 `comments` 를 수정하고, 가장 중요하게,

### **`comments` 를 사용하며 렌더링 하는 객체들은 모두 업데이트 돼요.**

따라서 해당 `comments` 를 인자로 받은 `Comment` 컴포넌트는 다시 렌더링이 됩니다. 좋아요 수가 증가된 형식을 반영된 채로요!

hyunju95 replied:
엇 질문 지웠는데 들켰네요 친절한설명 감솨합니다.
Comment 루프도는건 우선 석기님 말씀대로 api구축된 후 속도 보고 다시 논의해봐도 좋을것같네요~!

hyunju95 commented on src/Article/presentational/Comment.tsx:
```diff
@@ -0,0 +1,10 @@
+    </div>
+)
+
+export default CommentArea;
```
요 파일은 사용하지않는다면 delete 부탁드립니다~

Derek-94 replied:
엇 이게 올라갔네요.. 삭제하겠습니다! ㅎㅎ

hyunju95 commented on src/Article/presentational/CommentArea.tsx:
```diff
@@ -0,0 +1,87 @@
+    )
+}
+
+interface CommentAreaProps {
```
해당 interface는 apiProps로 들어가는게 좋을것같고 지금은 해당 파일이 src아래 있지만 Article을 위한 props들은 Article 하위에 위치시키는것이 어떨까 생각이드네여

Derek-94 replied:
네네 그럼 Article/ 폴더 내부에 argumentsProps.. 같은 폴더를 만들고, 그 내부에 `interface` 를 관리해볼까요?

hyunju95 replied:
넵 저는 그 방식이 좋을것같아요!

dayonxxi replied:
여기서 CommentAreaProps를 어디에 저장할지 얘기 나누시는 것 같아 대댓글 남깁니다! 현주님 말씀대로 진행하게 된다면 현재 apiProps에 있는 AuthorInfo와 ArticleDetailApi interface들도 Article폴더 내부로 들어가는거라고 생각하면 될까요??

Derek-94 replied:
@dayonxxi 음 음음 그건 따로 만들어야할 것 같아요!

1. Article 컴포넌트에서 쓰이는 `props` 들에 대한 `interface` 는 `Article` 폴더 내부에,
2. `API` 관련 `interface` 는 따로 `api` 폴더에 저장하는게 가장 좋을 것 같습니다!

hyunju95 commented on src/data/commentsInfo.json:
```diff
@@ -0,0 +1,26 @@
+    },
+    "content": "sfdsfsfsdfdsffsd",
+    "favorites": 2
+  }
```
제가 만든 json파일과 동일한 기능을 하는 석기님이만든 json파일이 중복되는거같은데 둘중 하나는 딜리트 시켜주시면 될것같아요

Derek-94 replied:
넵 이거 삭제하겠습니당 :)

hyunju95 reviewed (commented):
코드 내에 리뷰남겼습니다 확인부탁드려요
외롭게 오류 잡으시느라 고생하셨습니다ㅋㅋㅋ 🙀

dayonxxi commented on src/Article/presentational/CommentArea.tsx:
```diff
@@ -0,0 +1,87 @@
+                        <div className = "authorName"> {author.name} </div>  
+                    )}
+                </div>  
+                <div className = "contents"> {content} </div> 
```
제가 articlebody부분에서 본문내용을 표시할 때 {content}를 사용하는데 
혹시여기 {content}가 글의 본문내용과 해당 글의 댓글까지 담고 있는건가요??
제가 아직 이 부분이 제대로 이해되지 않아서 코멘트남깁니다!

Derek-94 replied:
아뇨! 다연님께서 본문내용 렌더링 할때 쓰이는 정보인 `dummyData` 의 `content`는 다른 정보입니다.

```typescript
export interface ArticleDetailApi {
    options: string[];
    title: string;
    anonymous: boolean;
    author: AuthorInfo;
    createdDate: string;
    content: string;
    hashtags: string[];
    favorites: number;
    wonders: number;
    clips: number;
}
```
위와 같이 생긴 `interface` 의 `content` 부분입니다. 즉, 본문의 내용만을 포함하고 있습니다!

dayonxxi reviewed (commented):
코드에 댓글로 리뷰 남겼습니다!!
수고 많으셨어요 :)

Derek-94 committed:
refactor: PR 리뷰받은 내용 적용.

- props로 넘겨지는 interface 모듈화.
- 쓰이지 않는 data/ 내부 .json 파일 삭제.
- 쓰이지 않는 파일 Comment.tsx 삭제.

Derek-94 reviewed (commented):
@hyunju95 님, @dayonxxi 님!

리뷰해주신 내용 중심으로 수정해서 다시 push 했습니다!

hyunju95 reviewed (approved):
수정사항 반영 확인완료했습니다~! 
수고하셨습니다! 
디벨롭에 머지되면 받아서 css작업 마무리할게요🤗
